pkg/middleware/batch.go

```diff
@@ -117,59 +128,86 @@ type BatchRequest struct {
-				if e != nil {
-					GetLogger(r).Error().Err(e).Msg("cannot convert operation body to bytes for operation id " + op.OperationID)
-					continue
-				}
-				reader := bytes.NewReader(bytesBody)
-				opReq, e := http.NewRequest(op.Method, op.URL, reader)
-				if e != nil {
-					GetLogger(r).Error().Err(e).Msg("cannot make a new request for operation id: " + op.OperationID)
-					continue
-				}
+	bytesBody, err := json.Marshal(op.Body)
+	if err != nil {
+		col.Status = http.StatusBadRequest
+		return col, fmt.Errorf("cannot convert operation body to bytes for operation id: %s with error: %v", op.OperationID, err)
+	}
+	reader := bytes.NewReader(bytesBody)
+	opReq, err := http.NewRequest(op.Method, op.URL, reader)
+	if err != nil {
+		col.Status = http.StatusBadRequest
+		return col, fmt.Errorf("cannot make a new request for operation id: %s, with error: %v", op.OperationID, err)
+	}
+
+	for headerKey, headerValue := range op.Headers {
+		opReq.Header.Add(headerKey, headerValue)
+		col.headerMap[headerKey] = []string{headerValue}
+	}
+
+	for paramKey, paramValue := range op.Params {
+		values := opReq.URL.Query()
+		values.Add(paramKey, paramValue)
+		opReq.URL.RawQuery = values.Encode()
+	}
+
+	next.ServeHTTP(&col, opReq)
+	col.EndedAt = time.Now()
+	return col, nil
+}
+
+// BatchRouter intercepts requests for the given url to return a StatusOK.
+func BatchRouter(batchRequests config.BatchRequestsConfig) func(http.Handler) http.Handler {
 
-				col := NewResponseCollector(op)
+	f := func(next http.Handler) http.Handler {
 
-				for headerKey, headerValue := range op.Headers {
-					opReq.Header.Add(headerKey, headerValue)
-					col.headerMap[headerKey] = []string{headerValue} // Not sure this is needed
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			if r.Method == "POST" && strings.HasSuffix(strings.ToLower(r.URL.Path), "/batch") {
+
+				decoder := json.NewDecoder(r.Body)
+				var req BatchRequest
+				err := decoder.Decode(&req)
+				if err != nil {
+					w.WriteHeader(http.StatusBadRequest)
+					render.JSON(w, r, `{"error": "cannot decode the operation body"}`)
+					return
 				}
 
-				for paramKey, paramValue := range op.Params {
-					values := opReq.URL.Query()
-					values.Add(paramKey, paramValue)
-					opReq.URL.RawQuery = values.Encode()
+				if len(req.Operations) > batchRequests.OperationsLimit {
+					GetLogger(r).Info().Msg(fmt.Sprintf("Too many operations, setting to the first %d operations", batchRequests.OperationsLimit))
+					req.Operations = req.Operations[:batchRequests.OperationsLimit]
 				}
 
-				next.ServeHTTP(&col, opReq)
-				col.EndedAt = time.Now()
+				batchRes := NewBatchResponse()
+				var eg errgroup.Group
 
-				// Append response item
-				batchRes.append(col)
-			}
+				ch := make(chan struct{}, batchRequests.ParallelRequests)
 
-			render.JSON(w, r, batchRes)
-			return
+				for _, op := range req.Operations {
+					op := op
+					eg.Go(func() error {
+						ch <- struct{}{}
```

Comment:
This approach starts _all_ the the goroutines, then blocks their execution. We should probably block the goroutine from being submitted until there is an available lease.

Reply:
yeah, starting all goroutines at once is cheap (go can handle a mln of go routines started at once, since they are green threads) , and I think it will be faster that way instead of starting goroutines in batches.

Reply:
Hm, ok. I think what makes this work in this scenario is that we're blocking on a channel. So the go scheduler knows it can put this goroutine into a waiting state. If the goroutine was blocked on some other system call, then I think this would have adverse implications. I'll probably do some research on this to satisfy my own understanding.

Reply:
after a much thought, I am changing it